cpp/src/arrow/engine/substrait/type_internal.cc

```diff
@@ -0,0 +1,434 @@
+                                                                 A&&... args) {
+  RETURN_NOT_OK(CheckVariation(type));
+
+  return std::make_pair(std::static_pointer_cast<DataType>(
+                            std::make_shared<ArrowType>(std::forward<A>(args)...)),
+                        IsNullable(type));
+}
+
+template <typename TypeMessage, typename... A>
+Result<std::pair<std::shared_ptr<DataType>, bool>> FromProtoImpl(
+    const TypeMessage& type, std::shared_ptr<DataType> type_factory(A...), A&&... args) {
+  RETURN_NOT_OK(CheckVariation(type));
+
+  return std::make_pair(
+      std::static_pointer_cast<DataType>(type_factory(std::forward<A>(args)...)),
+      IsNullable(type));
+}
+
+template <typename Types, typename Names = const std::string*>
+Result<FieldVector> FieldsFromProto(
+    int size, const Types& types,
+    const Names* names = static_cast<const Names*>(nullptr)) {
+  FieldVector fields(size);
+  for (int i = 0; i < size; ++i) {
+    ARROW_ASSIGN_OR_RAISE(auto type_nullable, FromProto(types[i]));
+
+    std::string name = names ? std::move((*names)[i]) : "";
+    fields[i] =
+        field(std::move(name), std::move(type_nullable.first), type_nullable.second);
+  }
+  return fields;
+}
+
+}  // namespace
+
+Result<std::pair<std::shared_ptr<DataType>, bool>> FromProto(const st::Type& type) {
+  switch (type.kind_case()) {
+    case st::Type::kBool:
+      return FromProtoImpl<BooleanType>(type.bool_());
+
+    case st::Type::kI8:
+      return FromProtoImpl<Int8Type>(type.i8());
+    case st::Type::kI16:
+      return FromProtoImpl<Int16Type>(type.i16());
+    case st::Type::kI32:
+      return FromProtoImpl<Int32Type>(type.i32());
+    case st::Type::kI64:
+      return FromProtoImpl<Int64Type>(type.i64());
+
+    case st::Type::kFp32:
+      return FromProtoImpl<FloatType>(type.fp32());
+    case st::Type::kFp64:
+      return FromProtoImpl<DoubleType>(type.fp64());
+
+    case st::Type::kString:
+      return FromProtoImpl<StringType>(type.string());
+    case st::Type::kBinary:
+      return FromProtoImpl<BinaryType>(type.binary());
+
+    case st::Type::kTimestamp:
+      return FromProtoImpl<TimestampType>(type.timestamp(), TimeUnit::MICRO);
+    case st::Type::kTimestampTz:
+      return FromProtoImpl<TimestampType>(type.timestamp_tz(), TimeUnit::MICRO,
+                                          TimestampTzTimezoneString());
+    case st::Type::kDate:
+      // FIXME
+      // Substrait uses uint32_t to store dates, and further restricts the allowed
+      // range of dates to [1000-01-01..9999-12-31]. Does this mean the value should
+      // be interpreted as an offset from 1000-01-01 instead of the epoch? Or should
+      // the value be signed instead?
+      // Furthermore, simple_logical_types.md states that the equivalent arrow type
+      // is Date64 (which measures milliseconds rather than days). Is that incorrect?
```

Comment:
The current substrait docs on dates are not fully specified and yes likely incorrect on this type.

Reply:
I'll say though that the specific unit is only specified for literals